### ISSUE TITLE
Allow incomplete split

### ISSUE THREAD

**Issue by danieltowner** [mattkretz/std-simd-feedback#31]
A simd can be broken down into pieces of specific sizes:

    const auto [a, b, c] = split<2, 3, 2>(simd_values);

At the moment split requires the sum of the pieces to match the size of the input value, but this seems overly restrictive. I can see that it helps guide correct behaviour by ensuring that nothing gets left behind, but I think it more likely that code might not know the size of the simd_value, and consequently it should only be an error if the number of pieces is too large. For example, in the example above perhaps `simd_value' is a native simd type. This might have enough elements to supply 2 + 3 + 2 simd pieces, with some padding left over, but computing how much padding is needed makes for an unnecessarily verbose call.

I propose that split should require the sum of the pieces to be no larger than the number of elements, rather than exactly the same size.

**Comment by danieltowner** [comment]
Also, rather than removing split from the overload set if the sum of sizes doesn't meet a constraint (i.e., exact size of input, or less than size in my proposal), then this should be a static_assert failure instead, to make it clear that the programmer has misused split. I think it likely that if the programmer got the sizes wrong then that is an error, rather than something that should be fixed by using an alternative overload.

**Comment by mattkretz** [comment]
I agree on both counts. I've written ridiculous code for generically splitting a `simd`.
I also think I agree that no SFINAE on `split` is needed. If anyone wants to know whether a split can work, look at the `size`...

**Comment by mattkretz** [comment]
There are three variants of `split`:
* `split<N0, N1, ...>(...)`, requires the sum of all `N.` to be `size`
* `split<V>(...)`, requires `size % V::size == 0`
* `split_by<N>(...)`, requires `size % N == 0`

After thinking more about it, I find it error-prone to "silently" drop data. Just like casts don't allow resizing, because it could be unintentional.

```c++
void f(simd<float> x) {
  // the developer worked with x.size == 4, so he wrote:
  auto [lo, hi] = split<2, 2>(x);
```

I'm not ready to propose weaker rules (yet). But I'd be open for an opt-in:

```c++
auto [a, b, c] = split<N0, N1, std::dynamic_extent>(x);
```
"dynamic extent" is misnamed, but we'd likely want to use the same numerical constant (`size_t` max). If `N0 + N1 == x.size` it would need to return something empty or a tuple of two values. The latter would make the use of structured bindings hard.

Thoughts?

**Comment by danieltowner** [comment]
I like the idea of the opt-in.

The experimental Intel implementation also has the notion of split_uneven, which generates a tuple of values of some given maximum size, and a remainder:

    split_uneven<4>(fixed_size_simd<float, 10>())
    // return std::tuple<fixed_size_simd<float,4>, fixed_size_simd<float,4>, fixed_size_simd<float,2>>

Useful for loop peeling. Or similarly, ``split_native`` returns pieces of simd<> (i.e., the native size) plus a remainder, which is useful for breaking a simd down into intrinsic-sized pieces.

**Comment by mattkretz** [comment]
So, opt-in via different name or via template parameter? The former is probably simpler, both in terms of discovering the function and how to use it, and in terms of implementation?

**Comment by danieltowner** [comment]
Yes, by name initially I think.

**Comment by mattkretz** [comment]
Defer to a later paper. This seems like a strict addition.

**Comment by mattkretz** [comment]
@danieltowner Do you agree that this is resolved via the replacement `simd_split` in P1928R6?
